src/fluree/db/api/query.cljc

```diff
@@ -152,183 +157,110 @@
+                          retracts (<? (json-ld-resp/flakes->res db cache compact fuel 1000000
+                                                                 {:wildcard? true, :depth 0}
+                                                                 0 retract-flakes))]
+
+                      ;; t is always positive for users
+                      (cond-> {:t (- (flake/t (first t-flakes)))}
+                        asserts (assoc :assert asserts)
+                        retracts (assoc :retract retracts))))))
+           (async/merge)
+           (async/into [])
+           (<?)
+           (sort-by :t)))))
 
 (defn get-history-pattern
   [history]
-  (let [subject (cond (util/subj-ident? history)
-                      [history]
-
-                      (sequential? history)
-                      (if (empty? history)
-                        (throw (ex-info (str "Please specify an subject for which to search history. Provided: " history)
-                                        {:status 400
-                                         :error  :db/invalid-query}))
-                        history)
-
-                      :else
-                      (throw (ex-info (str "History query not properly formatted. Provided: " history)
-                                      {:status 400
-                                       :error  :db/invalid-query})))
-        [s p o t] [(get subject 0) (get subject 1) (get subject 2) (get subject 3)]
-
+  (let [[s p o t]     [(get history 0) (get history 1) (get history 2) (get history 3)]
         [pattern idx] (cond
                         (not (nil? s))
-                        [subject :spot]
+                        [history :spot]
 
                         (and (nil? s) (not (nil? p)) (nil? o))
                         [[p s o t] :psot]
 
                         (and (nil? s) (not (nil? p)) (not (nil? o)))
-                        [[p o s t] :post]
-
-                        :else
-                        (throw (ex-info (str "History query not properly formatted. Must include at least an subject or predicate to query. Provided: " history)
-                                        {:status 400
-                                         :error  :db/invalid-query})))]
+                        [[p o s t] :post])]
     [pattern idx]))
 
+(def History
+  [:map {:registry {::iri [:or :keyword :string]
+                    ::context [:map-of :any :any]}}
+   [:history
+    [:orn
+     [:subject ::iri]
+     [:flake
+      [:or
+       [:catn
+        [:s ::iri]]
+       [:catn
+        [:s [:maybe ::iri]]
+        [:p ::iri]]
+       [:catn
+        [:s [:maybe ::iri]]
+        [:p ::iri]
+        [:o [:not :nil]]]]]]]
+   [:context {:optional true} ::context]
+   [:t {:optional true}
+    [:and
+     [:map
+      [:from {:optional true} pos-int?]
+      [:to {:optional true} pos-int?]]
+     [:fn (fn [{:keys [from to] :as x}] (or from to))]]]])
+
```

Comment:
I'm not suggestion this change at this point, just a good example here to express what I think would be more intuitive, faster, and provide a better user experience.

Why not just include a comment like:  
```
;; history query parameter is a map as follows:
{:history [s p] ;; where s is subject and optional p is predicate value for history results
 :t {:from 1  ;; optional - when provided will limit results to t values within range inclusive  
     :to   3}}
```
Which tells me more information than this proprietary DSL does - and doesn't require me to go learn a new library to decipher

Then, handle validation with a short function that is capable of providing more useful feedback to the user.

Reply:
I think a comment here is a good idea to clarify that admittedly unfamiliar malli schema. Malli has some really helpful facilities for [friendly error messages](https://github.com/metosin/malli#custom-error-messages) that we're not currently utilizing, we could leverage that to provide detailed, helpful error messages.

But there are some additional benefits to using malli here:
 - speed - malli is [as fast or faster](https://github.com/metosin/malli#performance) than vanilla clojure for this sort of validation and parsing
 - consolidated validation, so it's all done in one blow
 - declarative documentation of the data shapes - the actual rules for what sort of flake pattern is allowed is actually quite a bit more complex than just a vector, and writing manual verification of that takes quite a bit of space.

Reply:
To elaborate on the performance claims, I went and benchmarked the malli validator against a hand-rolled clojure one this morning:
```
(def history-query-validator
  (m/validator
    [:map {:registry {::iri [:or :keyword :string]
                      ::context [:map-of :any :any]}}
     [:history
      [:orn
       [:subject ::iri]
       [:flake
        [:or
         [:catn
          [:s ::iri]]
         [:catn
          [:s [:maybe ::iri]]
          [:p ::iri]]
         [:catn
          [:s [:maybe ::iri]]
          [:p ::iri]
          [:o [:not :nil]]]]]]]
     [:context {:optional true} ::context]
     [:t {:optional true}
      [:and
       [:map
        [:from {:optional true} pos-int?]
        [:to {:optional true} pos-int?]]
       [:fn {:error/message "Either \"from\" or \"to\" `t` keys must be provided."}
        (fn [{:keys [from to]}] (or from to))]
       [:fn {:error/message "\"from\" value must be less than or equal to \"to\" value."}
        (fn [{:keys [from to]}] (if (and from to)
                                  (<= from to)
                                  true))]]]]))
```
vs
```
(defn valid-history?
  [query]
  (let [{:keys [history t context]} query]
    (cond (and context (not (map? context)))
          :invalid/context
          (and t (set/subset? #{:from :to} (set (keys t)))
               (not-empty (filter pos-int? (vals t)))
               (if (and (:from t) (:to t))
                 (<= (:from t) (:to t))
                 true))

          (and t (or (:from t) (:to t)))
          {:error :t :message "Either \"from\" or \"to\" `t` keys must be provided."}
          (and t (not-empty (filter pos-int? (vals t))))
          {:error :t :message "Supplied \"from\" or \"to\" values must be positive integers."}
          (and t (if (and (:from t) (:to t))
                   (<= (:from t) (:to t))
                   true))
          {:error :t :message "Supplied \"from\" value must be less than or equal to \"to\" value."}

          (not history)
          {:error :history :message "Missing required key."}
          (or (or (string? history) (keyword? history))
              (and (sequential? history)
                   (or (and (= 1 (count history))
                            (not (nil? (first history)))
                            (or (string? (first history)) (keyword? (first history))))
                       (and (= 2 (count history))
                            (not (nil? (second history)))
                            (or (string? (second history)) (keyword? (second history))))
                       (and (= 3 (count history))
                            (not (nil? (second history)))
                            (or (string? (second history)) (keyword? (second history)))
                            (not (nil? (nth history 2))))))))))
```

The malli validator is almost twice the speed of the clojure one:
```
(require '[criterium.core :as criterium])
(require '[fluree.db.api.query :as query])

  (criterium/quick-bench (query/valid-history? {:history [:ex/dan :ex/x] :t {:from 1 :to 3}}))

 "Evaluation count : 610338 in 6 samples of 101723 calls.
             Execution time mean : 1.012344 µs
    Execution time std-deviation : 40.352968 ns
   Execution time lower quantile : 974.547438 ns ( 2.5%)
   Execution time upper quantile : 1.068644 µs (97.5%)
                   Overhead used : 5.883147 ns"

(criterium/quick-bench (query/history-query? {:history [:ex/dan :ex/x] :t {:from 1 :to 3}}))

  "Evaluation count : 984732 in 6 samples of 164122 calls.
             Execution time mean : 588.280715 ns
    Execution time std-deviation : 5.862140 ns
   Execution time lower quantile : 580.374794 ns ( 2.5%)
   Execution time upper quantile : 593.829236 ns (97.5%)
                   Overhead used : 5.883147 ns"
```
So between that and the other benefits that malli brings it seems like a straightforward win. It was designed to be lightweight on the browser as well, with some advanced tools for trimming everything that's not required out of the final artifact.

Reply:
Now that I'm looking at the validator, though - can we simplify the allowed :history forms?

Specifically, can we ditch the `{:history :my/subject-id}` form? That way you always have to supply a vector. We already support `{:history [:my/subject-id]}`, and I don't think the first form offers much syntax sugar over the flake form.

Reply:
I like using a purpose-build validation library for this. I think it's yet another example of [Greenspun's tenth rule](https://en.wikipedia.org/wiki/Greenspun%27s_tenth_rule). Whatever we hand roll will most likely be buggier and less capable then something good we can pull off the shelf. 

I do think we should standardize on how we do these validations though. I think we'd need a very good reason to use two different validation libraries, for example. Because Malli is both _very_ fast and has friendly error messages built in, I think it's probably the best choice for us.

Reply:
The logic to determine what is valid still must be done. e.g. is {:history [:ex/brian]} valid? Maybe, if a keyword context is used, and if the compacted form expands, and if the DB knows of this existence.

Is {:t {:from 42}} valid? Maybe, if the ledger's t is >=42.

To provide useful error messages those steps need to be done and good "user-friendly" feedback can be given depending on the failure point.

Is it possible someone will send {:t {:from "foo"}}? Yes... and I guess I'll be glad we had Mali/Spec/whatever to tell us that when it happens 1 of every billion times this situation arises, even though two steps in we would have discovered the issue anyhow.

If we were doing data shape formatting because this data was passed into a DB then I'd totally say this tool makes sense. We wouldn't be picking the input apart, we are a proxy but must make sure it is "right". That high value use case just seems misplaced here.

That, and of course we already *are* the downstream system for true data inputs, and have SHACL for shape validation. I'd at least be happier if we used our own code whose job it is to do this for data.

Reply:
I think I understand what you mean - some validation requires additional inputs besides just what was submitted (for example, that the submitted subject actually exists in the db). So there's two steps of validation: 1) validation that the input by itself is valid and 2) validation that the input is coherent with the db. 

I think we always have to do 1), even if it only happens every once in a billion times. And it's best to do it as fast as possible, which where malli really helps.

Reply:
I think you are doing the validation anyhow by the nature of this input needing to be picked apart and evaluated. You are using a data validation tool to validate an instruction set, which still needs to be processed (and by nature, validated) anyhow.

I think this is a great tool for taking data inputs and passing them to other systems (like a DB) where I'd otherwise have to write custom validation code. 90% of APIs do this, and probably 90% of them should use this.. it offers great value.

This is not that use case.  In this use case, using this tool means we now have the same validation (hopefully!) being done twice, in two different parts of the code. In addition we've created a new dependency, a new "language" someone must learn to contribute, and slowed down the system because it is doing the same work twice.

If we gained something worth the cost then great... I'm having a hard time seeing what we gain here. Not documentation, as by the nature of our tool we have much more detailed doc on this input in our end-user documentation already. Not testing, because any issue this would catch would be caught anyway by our processing, but importantly the reverse is not true... the processing will catch errors this would never catch. Maybe I'm missing some other benefit... but we are paying a high price to have it, so the benefit should be equally high.